File: Backend/src/entities/Quarto.ts
/**
 * Entity: Quarto
 * Responsabilidade: validação pura das regras de negócio.
 * Nunca toca o banco de dados.
 */

export interface QuartoItemInput {
  catalogo_id: number;
  quantidade:  number;
}

export interface CreateQuartoInput {
  numero:              string;
  categoria_quarto_id: number;
  descricao?:          string | null;
  valor_override?:     number | null;
  disponivel?:         boolean;
  itens?:              QuartoItemInput[];
}

export interface UpdateQuartoInput {
  numero?:              string;
  categoria_quarto_id?: number;
  descricao?:           string | null;
  valor_override?:      number | null;
  disponivel?:          boolean;
  itens?:               QuartoItemInput[];
}

export interface QuartoItemSafe {
  catalogo_id: number;
  nome:        string;
  categoria:   string;
  quantidade:  number;
}

export interface QuartoSafe {
  id:                  number;
  numero:              string;
  categoria_quarto_id: number;
  disponivel:          boolean;
  descricao:           string | null;
  valor_override:      string | null;  // DECIMAL retorna como string pelo driver pg
  itens:               QuartoItemSafe[];
}

export class Quarto {
  private static validateNumero(numero: unknown): string {
    if (typeof numero !== 'string' || numero.trim().length === 0 || numero.trim().length > 10)
      throw new Error('numero inválido: deve ter entre 1 e 10 caracteres');
    return numero.trim();
  }

  private static validateCategoriaId(id: unknown): number {
    const v = Number(id);
    if (!Number.isInteger(v) || v <= 0)
      throw new Error('categoria_quarto_id inválido: deve ser um inteiro maior que zero');
    return v;
  }

  private static validateDescricao(desc: unknown): string | null {
    if (desc === null || desc === undefined) return null;
    if (typeof desc !== 'string' || desc.length > 500)
      throw new Error('descricao inválida: deve ter no máximo 500 caracteres');
    return desc.trim() || null;
  }

  private static validateValorOverride(valor: unknown): number | null {
    if (valor === null || valor === undefined) return null;
    const v = Number(valor);
    if (isNaN(v) || v <= 0)
      throw new Error('valor_override inválido: deve ser um número maior que zero');
    return v;
  }

  private static validateItens(itens: unknown): QuartoItemInput[] {
    if (!Array.isArray(itens)) throw new Error('itens deve ser um array');
    return itens.map((item, i) => {
      const catalogoId = Number((item as any).catalogo_id);
      if (!Number.isInteger(catalogoId) || catalogoId <= 0)
        throw new Error(`itens[${i}].catalogo_id inválido: deve ser um inteiro maior que zero`);

      const quantidade = (item as any).quantidade !== undefined ? Number((item as any).quantidade) : 1;
      if (!Number.isInteger(quantidade) || quantidade <= 0)
        throw new Error(`itens[${i}].quantidade inválida: deve ser um inteiro maior que zero`);

      return { catalogo_id: catalogoId, quantidade };
    });
  }

  /** Valida campos obrigatórios para criação. */
  static validate(input: unknown): CreateQuartoInput {
    const data = input as Record<string, unknown>;

    const result: CreateQuartoInput = {
      numero:              this.validateNumero(data.numero),
      categoria_quarto_id: this.validateCategoriaId(data.categoria_quarto_id),
    };

    if (data.descricao      !== undefined) result.descricao      = this.validateDescricao(data.descricao);
    if (data.valor_override !== undefined) result.valor_override = this.validateValorOverride(data.valor_override);
    if (data.disponivel     !== undefined) {
      if (typeof data.disponivel !== 'boolean') throw new Error('disponivel deve ser true ou false');
      result.disponivel = data.disponivel;
    }
    if (data.itens !== undefined) result.itens = this.validateItens(data.itens);

    return result;
  }

  /** Valida campos presentes para atualização parcial. */
  static validatePartial(input: unknown): UpdateQuartoInput {
    const data = input as Record<string, unknown>;
    const result: UpdateQuartoInput = {};

    if (data.numero              !== undefined) result.numero              = this.validateNumero(data.numero);
    if (data.categoria_quarto_id !== undefined) result.categoria_quarto_id = this.validateCategoriaId(data.categoria_quarto_id);
    if (data.descricao           !== undefined) result.descricao           = this.validateDescricao(data.descricao);
    if (data.valor_override      !== undefined) result.valor_override      = this.validateValorOverride(data.valor_override);
    if (data.disponivel          !== undefined) {
      if (typeof data.disponivel !== 'boolean') throw new Error('disponivel deve ser true ou false');
      result.disponivel = data.disponivel;
    }
    if (data.itens !== undefined) result.itens = this.validateItens(data.itens);

    if (!Object.keys(result).length) throw new Error('Nenhum campo para atualizar');
    return result;
  }
}

File: Frontend/lib/features/rooms/domain/models/quarto.dart
class QuartoItemModel {
  final int catalogoId;
  final String nome;
  final String categoria;
  final int quantidade;

  const QuartoItemModel({
    required this.catalogoId,
    required this.nome,
    required this.categoria,
    required this.quantidade,
  });

  factory QuartoItemModel.fromJson(Map<String, dynamic> json) {
    return QuartoItemModel(
      catalogoId: json['catalogo_id'] as int? ?? 0,
      nome: json['nome'] as String? ?? '',
      categoria: json['categoria'] as String? ?? '',
      quantidade: json['quantidade'] as int? ?? 1,
    );
  }
}

class QuartoModel {
  final int id;
  final String numero;
  final int categoriaQuartoId;
  final bool disponivel;
  final String? descricao;
  final double? valorDiaria;
  final List<QuartoItemModel> itens;

  const QuartoModel({
    required this.id,
    required this.numero,
    required this.categoriaQuartoId,
    required this.disponivel,
    this.descricao,
    this.valorDiaria,
    this.itens = const [],
  });

  factory QuartoModel.fromJson(Map<String, dynamic> json) {
    final itensRaw = json['itens'] as List<dynamic>? ?? [];
    return QuartoModel(
      id: json['id'] as int? ?? 0,
      numero: json['numero'] as String? ?? '',
      categoriaQuartoId: json['categoria_quarto_id'] as int? ?? 0,
      disponivel: json['disponivel'] as bool? ?? true,
      descricao: json['descricao'] as String?,
      valorDiaria: _parseDouble(json['valor_diaria']),
      itens: itensRaw
          .map((i) => QuartoItemModel.fromJson(i as Map<String, dynamic>))
          .toList(),
    );
  }

  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }
}
